src/olympia/addons/fields.py

```diff
@@ -214,15 +214,17 @@ def to_internal_value(self, data):
         try:
             _, ext = os.path.splitext(data.name)
             if ext == '.zip':
-                # testzip() returns None if there are no broken CRCs.
+                # For zip files, opening them though SafeZip() checks that we
+                # can accept the file and testzip() on top of that returns None
+                # if there are no broken CRCs.
                 if SafeZip(data).zip_file.testzip() is not None:
                     raise zipfile.BadZipFile()
             else:
-                # For tar files we need to do a little more work.
+                # For tar files, opening them through SafeTar() checks that we
+                # can accept it.
                 mode = 'r:bz2' if ext == '.bz2' else 'r:gz'
-                with tarfile.open(mode=mode, fileobj=data) as archive:
-                    for member in archive.getmembers():
-                        archive_member_validator(archive, member)
+                with SafeTar.open(mode=mode, fileobj=data):
+                    pass
```

Comment:
Is it possible (practical) to have a similar method for validating for `SafeTar` and `SafeZip`?  (i.e. not `if SafeZip(data).zip_file.testzip() is not None` for zips and an entirely different `with SafeTar.open(mode=mode, fileobj=data)` for tars.)

Reply:
Possibly - we could do the `testzip()` on opening for all zip files, but we'd need to add more code to ignore any errors about that in the scanners code, since we might already have such broken zip files in storage right now...

Reply:
I looked at this and I'm not convinced it's a good idea to do that everytime we use `SafeZip`, as it forces us to read through all the contents of all the files in the archive. 

At add-on upload, if a developer uploads a broken archive, their extension won't work, so it's their responsibility, so we don't bother. We check at source code upload because they might not realize then that easily.

Reply:
Both `SafeZip` and `SafeTar` perform cheap validation checks on init (stuff around filenames).

`testzip()` is a special thing added on top to check CRC, just for source code uploads: it goes through all the files inside the archive and tries to read them, which takes some time & memory to do.

Reply:
"For tar files, opening them through SafeTar.open() checks that we"